diff --git a/argos@pew.worldwidemann.com/button.js b/argos@pew.worldwidemann.com/button.js
index 51f0101..71ff806 100644
--- a/argos@pew.worldwidemann.com/button.js
+++ b/argos@pew.worldwidemann.com/button.js
@@ -32,7 +32,7 @@ class ArgosButton extends PanelMenu.Button {
     this._updateInterval = settings.updateInterval;
 
     this._lineView = new ArgosLineView();
-    this._lineView.setMarkup("<small><i>" + GLib.markup_escape_text(file.get_basename(), -1) + " ...</i></small>");
+    this._lineView.setMarkup("<small><i>...</i></small>");
     Utilities.getActor(this).add_child(this._lineView);
 
     this._isDestroyed = false;
