diff --git a/src/filewatcher.js b/src/filewatcher.js
index 4af7eb8..caa734f 100644
--- a/src/filewatcher.js
+++ b/src/filewatcher.js
@@ -27,6 +27,7 @@ export function probeDirectories() {
         `${GLib.get_user_config_dir()}/syncthing`,
         `${GLib.get_home_dir()}/snap/syncthing/common/syncthing`,
         `${GLib.get_home_dir()}/.var/app/me.kozec.syncthingtk/config/syncthing`,
+        `/var/lib/syncthing/.config/syncthing`
     ];
     for (let dir of directories) {
         let configfile = Gio.File.new_for_path(`${dir}/config.xml`);
